### PR TITLE
verify a contract

### PR DESCRIPTION
**Pull request by r-gochain**
```
./web3 -testnet contract deploy ERC20.bin
Contract has been successfully deployed with transaction: 0x0c308984154a8029d7fad8d53057719e44a2a79d8d8bd62ddc950f4a9502ce20
Contract address is: 0x6F8c9989250F81ac93e3f38e57701D261fD594DD

./web3 contract verify --explorer-api https://testnet-explorer.gochain.io/api --address 0x6F8c9989250F81ac93e3f38e57701D261fD594DD  --contract-name ERC20 ERC20Burnable_flatten.sol
Your contract is successfully verified
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -316,6 +316,20 @@ web3 contract build FILENAME.sol --solc-version SOLC_VERSION
...
+```sh
+web3 contract flatten FILENAME.sol -o OUTPUT_FILE
+```
+
+**Parameters:**
+
+* FILENAME - the name of the .sol file, eg: `hello.sol`
+
+* OUTPUT_FILE (optional) - the output file
+
```
Comment (jmank88) [inline README.md]: Left over?

**Comment by jmank88** [comment]
| ./web3 contract verify --explorer-api https://testnet-explorer.gochain.io/api ...

Can we derive the `explorer-api` from the `network` so that it is only necessary when using a custom url?

**Comment by jmank88** [comment]
What about adding a `-verify` flag to `contract deploy`? Or maybe that is weird since deploy only takes the binary?

File: README.md (hunk 1 above)
Comment (r-gochain) [reply README.md]: yep, forgot to commit last time

**Comment by r-gochain** [comment]
> What about adding a `-verify` flag to `contract deploy`? Or maybe that is weird since deploy only takes the binary?

I thought about that, even one command - build-deploy-verify, but each step requires own set of parameters and verify  also requires explorer-api url, so I believe it's better to keep it a separate one

**Comment by treeder** [comment]
This is kind of bugging me that it's not really user friendly. They have to generate a separate flattened contract to verify?

**Comment by jmank88** [comment]
| This is kind of bugging me that it's not really user friendly. They have to generate a separate flattened contract to verify?

I think it's a little weird because flattening is naturally part of the binary creation, but here we are doing it afterwards but earlier in the process on just the source files. Could we rework the ux a bit, like by having build output both the binary and the flattened source at the same time? And then deploy could `--verify` optionally, and already have the flat source on hand?

**Comment by r-gochain** [comment]
I was going to make flattening a mandatory part before building binary (because it's simply doesn't make situation worse - single files are processed in the same way and files with imports going to be flattened)

**Comment by r-gochain** [comment]
I'll review the UX

**Comment by r-gochain** [comment]
@jmank88 @treeder updated a bit (now always flattening), but it's not very convenient to verify after deploy because exporer cannot verify a contract  before scanning it (usually it takes a couple of minutes), so deploy with verify would be running for 3-4 minutes which is not the best experience for user

**Comment by jmank88** [comment]
Can explorer check the contract in real time via RPC (and update the DB) when it's not yet loaded? Or could that clash with the separate grabber process?

**Comment by r-gochain** [comment]
Grabber uses upserts so shouldn't meet any issues, in theory it could work, let me try this

**Comment by r-gochain** [comment]
```./web3 -testnet contract deploy --explorer-api http://localhost:8080/api --verify ERC20Pausable_flatten.sol ERC20.bin
Contract has been successfully deployed with transaction: 0x0f0d1a4282b3e570428c241be3eec1b17aafdca6d8eb621c3d082c9870e6c8c0
Contract address is: 0x7951fF68A0Ac9D4dF2B93654acAa931939fb6078
Your contract is successfully verified!
```

**Comment by r-gochain** [comment]
related PR - https://github.com/gochain/explorer/pull/438

**Comment by r-gochain** [comment]
@jmank88 @treeder PTAL (explorer PR has been merged so I'll deploy and rollout shortly)

File: README.md (hunk 1 above)
Comment (jmank88) [reply README.md]: Shouldn't this be removed?

File: README.md [hunk 2]
```diff
@@ -397,7 +416,20 @@ web3 transaction ADDRESS_HASH
...
+### Verify a smart contract to a block explorer
+
+```sh
+web3 contract verify --explorer-api EXPLORER_API_URL --address CONTRACT_ADDRESS  --contract-name CONTRACT_NAME FILENAME.sol
+```
+
+**Parameters:**
+
+* EXPLORER_API_URL - url to block explorer API (eg https://testnet-explorer.gochain.io/api), you could also use -testnet instead
```
Comment (jmank88) [inline README.md]:
```suggestion
* EXPLORER_API_URL - URL for block explorer API (eg https://testnet-explorer.gochain.io/api) - Optional for GoChain networks, which use `{testnet-}explorer.gochain.io` by default.
```

File: cmd/web3/main.go [hunk 3]
```diff
@@ -259,6 +259,45 @@ func main() {
 							Usage:       "Allow contract to be upgraded",
 							Destination: &upgradeable,
 							Hidden:      false},
+						cli.StringFlag{
+							Name:  "verify",
+							Usage: "Source code of the contract",
+						},
+						cli.StringFlag{
+							Name:  "explorer-api",
+							Usage: "Explorer API URL",
```
Comment (jmank88) [inline cmd/web3/main.go]:
```suggestion
							Usage: "Explorer API URL. Optional for GoChain networks, which use `{testnet-}explorer.gochain.io` by default",
```

**Comment by r-gochain** [comment]
@jmank88 one more time?(I resolved comments but cannot merge)

**Comment by jmank88** [comment]
There is a still a doc in the readme for a 'flatten' command.

**Comment by r-gochain** [comment]
@jmank88 I believe it still should be there to cover the case when user wants to verify contract with imports that he deployed before - just for convenience

**Pull request merged**